### ISSUE TITLE
Select first IP instead of last when X-Forwarded-For is activated.

### ISSUE THREAD

**Issue by fernandezjose** [divolte/divolte-collector#177]
@sladestewart2

We have a business requirement where we need to collect the first IP in the collection of IPs (if a collection is present) instead of the last one (as is the default behavior in the ProxyAdjacentPeerAddress). 

We know how to achieve this in different ways since the changes are really straightforward, however, we wanted to reach out to you first and have suggestions from you on how you would like us to approach this issue.

The idea is to add a new property next to `divolte.global.server.use_x_forwarded_for` where we would allow users to specify if they want the first or last IP in the collection. We will default it to last so that your current implementation is not affected.

The property would be something like `divolte.global.server.select_first_ip` default to false.

Our first and simpler approach would be as simple as "copying" `ProxyAdjacentPeerAddress` to a new handler `ProxyDistantPeerAddress` and modify so that it reads the first ip instead of the last one. The decision making wether to use one handler or the other would occur in the Server.java class. This is the quicker approach but there is some repetitive code/functionality.

Second approach is to create a "complementary" handler which will strip down the collection of IPs to one, either first or last based on the property, and then pass that request (next.execute....) to `ProxyAdjacentPeerAddress`. This approach would feel seamless since the handler would be passed as one more in the chain.

Third approach would be to pass the new property as an argument in the constructor of `ProxyAdjacentPeerAddress` and make that determination of picking the first or last ip internally in that handler. This approach would require to modify the existing handler in order to accommodate the new property but seems the ideal way to achieve this given the little amount of work.

Thank you!

**Comment by friso** [comment]
In your mapping, you can use `map header('X-Forwarded-For').first() onto 'remoteIpAddress'` and then map that value onto a field in your schema. Would that work for you?

Also note that if you are using ip2geo, you can use the overload that takes a alternative remote ip value producer instead of the built in remote host. So, `ip2geo(remoteIpAddress)`, would use the value previously assigned from the first value of the header.

The addition of a config option for this case is of course possible, but there could be multiple other cases of requirements that we don't know about and eventually it might pollute the config a bit if we were to add them all.

**Comment by ppcuban** [comment]
@friso I wasn't aware of this option. I will try that instead.

**Comment by asnare** [comment]
There's also a problem with the current implementation: it's an all-or-nothing flag, which often isn't suitable. We really need a number which is how many hops to trust, defaulting to 0.

The current flag is equivalent to either 0 or ∞, when quite often you want a value of 1, 2, or 3. This is because most cloud-based load-balancers don't allow header scrubbing, and without that at the edge it's impossible to figure out the client IP without allowing the client to spoof it.

**Comment by asnare** [comment]
No, it's not really what I meant. Typically Divolte is X hops from an edge load-balancer, and that's the one whose `x-forwarded-for` header yields the 'right' remote IP address. Anything beyond is untrustworthy and anything nearer-by is internal HTTP routing to reach Divolte.

**Comment by friso** [comment]
Personally, I would try to solve this outside of Divolte closest to the edge. If you control your edge load balancer / reverse proxy, have that set a different header containing only the remote IP of the connection (e.g. `YourOrg-Source-Ip`; note that the use of the `X-` prefix is [deprecated](https://tools.ietf.org/html/rfc6648)).

Unfortunately, most cloud load balancing solutions don't do this and just use `X-Forwarded-For`. I think Divolte chooses the rightmost header, because Amazon's ELB [appends the client IP](http://docs.aws.amazon.com/elasticloadbalancing/latest/classic/x-forwarded-headers.html#x-forwarded-for) and nothing else. Google's cloud load balancer [appends one additional internal IP after the source IP](https://cloud.google.com/compute/docs/load-balancing/http/). For this reason configuring a position in the chain as a number from right as @asnare proposes would be a general solution.

`X-Forwarded-For` is not a standard of course, so there is no strict specification. There was an attempt to create [a spec for forwarding](https://tools.ietf.org/html/draft-ietf-appsawg-http-forwarded-10), but so far it doesn't look like anyone chose to implement it.

**Comment by ddehghan** [comment]
This can also happen when you have 2 load balancers. It is common that people use AWS ELB to offload SSL and then and other ELB to route the traffic to different micro services inside your cluster.

**Comment by friso** [comment]
@ddehghan Yeah, exactly. When that happens it's generally preferable to have your first internal ELB offload the `X-Forwarded-For` set by the ELB and set a private header with only a single IP instead. You can then use this private header in Divolte mapping regardless of the number of hops that your http makes internally.

**Comment by ddehghan** [comment]
The problem is AWS ELB (classic) is a piece of garbage and has no way of telling it to create custom headers. It seems easier to just get the first IP address of the current string here in the mapping file than trying to reconfigure the network and NLB. 


```
def ip = header('x-forwarded-for').first()
map ip onto 'xForwardedFor'
def ips = ip.split(',')
map ips[0] onto 'remoteHost'


This is content of the xForwardedFor in the log files.
{"xForwardedFor": "24.19.104.247, 100.96.2.1"}


```

I though something like the above code would work but after 4hrs i still can't get it to work. It should be so easy. It seems that ip variable is not a string. And I can't convert it to a string either so that I can have the split method. 


If someone can give me the code for how to extract the first IP address from that string it would be really nice.

**Comment by asnare** [comment]
There are 2 things going on here:

 - It can be confusing, but the mapping is declaring _what_ to map, not _how_ to map it. As a consequence you don't get to (easily) supply the code that will run during the mapping. The type of `ip` is not a string, which is why you can't split it. (Instead it's a handle to some code that will extract the first `x-forwarded-header` when mapping an event.)
 - The semantics of multiple headers. This is more complicated…

The HTTP 1.1 spec basically says that two requests with the following headers are semantically equivalent:
```
x-forwarded-for: 24.19.104.247
x-forwarded-for: 100.96.2.1
```
and
```
x-forwarded-for: 24.19.104.247, 100.96.2.1
```
At present `.first()` is either unaware of this or ambiguous, depending on how you look at it. If upstream delivered the former, you'd be fine without needing to split. But it's delivering the latter, which doesn't work. Upstream is allowed to do either.

Getting back to your problem: 'split' is a such a common thing that we should support it directly.

In the meantime I think you can implement what you want using the support for [regular expressions](http://divolte-releases.s3-website-eu-west-1.amazonaws.com/divolte-collector/0.6.0/userdoc/html/mapping_reference.html#regular-expression-matching).

**Comment by ddehghan** [comment]
Thanks @asnare and @friso 

My setup should be fairly common. I have AWS ELB (Classic) going to into Kubernetes. The most common way of deploying Kubernetes on AWS is with kops which creates an ELB (classic) for me automatically and that is the only AES ELB that they support at the moment which is an extremely limited load balancer. So I can't reconfigure the ELB at the moment. 

I tried the regex way but still I get null in xForwardedFor. Divolte docs says that .first() should return a string. But still no lock.

```
    def ip = header('x-forwarded-for').first()
    map ip onto 'remoteHost'

    def pathMatcher = match '(.*),' against ip
    when pathMatcher.matches() apply {
      map pathMatcher.group(1) onto 'xForwardedFor'
    }
```

**Comment by friso** [comment]
@ddehghan I think headers are case sensitive. Could you try `X-Forwarded-For`?

**Comment by ddehghan** [comment]
I checked the header the header it was not case sensitive. 

Unfortunately had to give up on divolte for now and implement my own backend. I could not take a dependency on a language and platform that I didn't know and was very hard for me to debug. So I implemented everything quickly in node. Hopefully at some point i come back to this and use what you guys have made. Thanks for the hard work. Something like this solution really needs to exists.
